pyfakefs is incompatible with pytest

This is a rather specific issue. To reproduce this, you need to have a failing doctest that uses Patcher.

Example doctest:
```
def foo():
    """
    >>> getfixture('fs')
    >>> foo()
    True
    """
    return False
```

Here's my `fs` fixture:
```
from fake_filesystem_unittest import _Patcher

_Patcher.SKIPNAMES.add('py')  # Ignore pytest components when faking filesystem


@pytest.yield_fixture
def fs():
    """ Fake filesystem. """
    patcher = _Patcher()
    patcher.setUp()
    yield
    patcher.tearDown()
```

Traceback: [pastebin](http://pastebin.com/raw.php?i=WuWj1sck)

Traceback after altering `_Patcher._findModules to skip literally everything: [pastebin](http://pastebin.com/raw.php?i=RaQjLhv2)

I'm not sure whether this error lies within pyfakefs, or can be resolved my adding some pytest submodules to `_Patcher.SKIPNAMES` set.

I tested this under Windows with the 2 tests that should crash:
```
C:\..\pyfakefs>py.test --doctest-modules pytest_doctest_test.py
============================= test session starts =============================
platform win32 -- Python 2.7.11, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: C:\..\pyfakefs, inifile:
collected 2 items

pytest_doctest_test.py FF

================================== FAILURES ===================================
____________________ [doctest] pytest_doctest_test.crashes ____________________
029         >>> open(name).read() == content
Expected:
    False
Got:
    True

C:\..\pyfakefs\pytest_doctest_test.py:29: DocTestFailure
__________________ [doctest] pytest_doctest_test.crashes_too __________________
028         >>> ) SyntaxError(name, content)
UNEXPECTED EXCEPTION: SyntaxError('invalid syntax', ('<doctest pytest_doctest_test.crashes_too[3]>', 1, 1, ') SyntaxError(name, content)\n'))
Traceback (most recent call last):

  File "c:\python27\Lib\doctest.py", line 1315, in __run

  File "<doctest pytest_doctest_test.crashes_too[3]>", line 1

    ) SyntaxError(name, content)

    ^

SyntaxError: invalid syntax

C:\..\pyfakefs\pytest_doctest_test.py:28: UnexpectedException
========================== 2 failed in 0.11 seconds ===========================
```
and it looks fine to me (tests fail as expected). Did I do something wrong, or is it a Linux-only problem, or maybe it is gone with the current versions of pyfakefs and pytest?
@jmcgeheeiv: Could you please check if the issue persists under Linux in case I did the test right?

@megapctr, could you comment on what is wrong with our results?

Result on OSX:
```
py.test --doctest-modules pytest_doctest_test.py
================================================== test session starts ==================================================
platform darwin -- Python 3.4.1 -- py-1.4.25 -- pytest-2.6.3
collected 3 items 

pytest_doctest_test.py F..

======================================================= FAILURES ========================================================
__________________________________________ [doctest] pytest_doctest_test.fails __________________________________________
024         """.format(
025         command="getfixture('fs')" if fake else "pass",
026         func_name=func_name,
027         result=result)
028 
029     return make_file
Expected:
    False
Got:
    True

/Users/johnm/git/pyfakefs/pytest_doctest_test.py:29: DocTestFailure
========================================== 1 failed, 2 passed in 0.03 seconds ===========================================
```
Result on Linux:
```
Forthcoming momentarily.
```

--- a/README.md
+++ b/README.md
@@ -51,7 +51,9 @@ See [Travis-CI](http://travis-ci.org) for
 ## Installation
 
 ### Compatibility
-pyfakefs works with Python 2.6 and above.  pyfakefs requires [mox3](https://pypi.python.org/pypi/mox3).
+pyfakefs works with Python 2.6 and above, on Linux and Windows.  pyfakefs requires [mox3](https://pypi.python.org/pypi/mox3).
+
+When using pyfakefs with [PyTest](doc.pytest.org), Pytest version 2.8.6 or above is required.
 
 ### PyPi
 [pyfakefs is available on PyPi](https://pypi.python.org/pypi/pyfakefs/).

--- a/conftest.py
+++ b/conftest.py
@@ -1,14 +1,15 @@
 import py
 import pytest
-from fake_filesystem_unittest import _Patcher
+from pyfakefs.fake_filesystem_unittest import Patcher
 
-_Patcher.SKIPMODULES.add(py)  # Ignore pytest components when faking filesystem
+
+Patcher.SKIPMODULES.add(py)  # Ignore pytest components when faking filesystem
 
 
 @pytest.yield_fixture
 def fs():
     """ Fake filesystem. """
-    patcher = _Patcher()
+    patcher = Patcher()
     patcher.setUp()
     yield
     patcher.tearDown()

--- a/pyfakefs/fake_filesystem.py
+++ b/pyfakefs/fake_filesystem.py
@@ -2818,6 +2818,7 @@ class FakeFileOpen(object):
     Returns FakeFile objects in a FakeFilesystem in place of the file()
     or open() function.
     """
+    __name__ = 'FakeFileOpen'
 
     def __init__(self, filesystem, delete_on_close=False, use_io=False):
         """init.

--- a/pytest_doctest_test.py
+++ b/pytest_doctest_test.py
@@ -37,13 +37,12 @@ def make_file_factory(func_name, fake, result):
 passes = make_file_factory('passes', fake=False, result=True)
 passes_too = make_file_factory('passes_too', fake=True, result=True)
 
-# Actually, it crashes if we try to remove the file :/
 passes_too.__doc__ = passes_too.__doc__.replace('>>> os.remove(name)',
                                                 '>>> pass')
 
 fails = make_file_factory('fails', fake=False, result=False)
 
-# Pytest raises an internal error when running these doctests:
+# Pytest versions below 2.8.6 raise an internal error when running these doctests:
 crashes = make_file_factory('crashes', fake=True, result=False)
 crashes_too = make_file_factory(') SyntaxError', fake=True, result=False)
 

--- a/pytest_doctest_test.py
+++ b/pytest_doctest_test.py
@@ -1,6 +1,10 @@
 """
+This is a test case for pyfakefs issue #45.
+
 To run these doctests, install pytest and run:
-$ py.test --doctest-modules pytest_doctest_test.py
+
+    $ py.test --doctest-modules pytest_doctest_test.py
+
 Add `-s` option to enable print statements.
 """
 from __future__ import unicode_literals
@@ -38,6 +42,7 @@ passes_too.__doc__ = passes_too.__doc__.replace('>>> os.remove(name)',
 
 fails = make_file_factory('fails', fake=False, result=False)
 
-# Pytest crashes when running these doctests:
-# crashes = make_file_factory('crashes', fake=True, result=False)
-# crashes_too = make_file_factory(') SyntaxError', fake=True, result=False)
+# Pytest raises an internal error when running these doctests:
+crashes = make_file_factory('crashes', fake=True, result=False)
+crashes_too = make_file_factory(') SyntaxError', fake=True, result=False)
+

--- a/pytest_doctest_test.py
+++ b/pytest_doctest_test.py
@@ -1,5 +1,6 @@
 """
 This is a test case for pyfakefs issue #45.
+This problem is resolved by using PyTest version 2.8.6 or above.
 
 To run these doctests, install pytest and run:
 

